webrtc/README.md

```diff
@@ -0,0 +1,304 @@
+  a host only advertise a single address, namely the host address, which is
+  assumed to be public.
+
+- Do the major (Go / Rust / ...) WebRTC implementations allow us to accept a
+  WebRTC connection from a remote node without previously receiving an SDP
+  packet from such host?
+
+- Can _Browser_ generate a _valid_ SDP packet for the remote node based on the
+  remote's Multiaddr, where that Multiaddr contains the IP, UDP port and TLS
+  certificate fingerprint (e.g.
+  `/ip6/2001:db8::/udp/1234/webrtc/certhash/<hash>/p2p/<peer-id>`)? _Valid_ in
+  the sense that this generated SDP packet can then be used to establish a
+  WebRTC connection to the remote.
+
+  Yes.
+
+### Browser to Browser
+
+Scenario: Browser _A_ wants to connect to Browser node _B_ with the help of
+server node _R_.
+
+- Replace STUN with libp2p's identify and AutoNAT
+  - https://github.com/libp2p/specs/tree/master/identify
+  - https://github.com/libp2p/specs/blob/master/autonat/README.md
+- Replace TURN with libp2p's Circuit Relay v2
+  - https://github.com/libp2p/specs/blob/master/relay/circuit-v2.md
+- Use DCUtR over Circuit Relay v2 to transmit SDP information
+  1. Transform ICE candidates in SDP to multiaddresses.
+  2. Transmit the set of multiaddresses to the remote via DCUtR.
+  3. Transform the set of multiaddresses back to the remotes SDP.
+  4. https://github.com/libp2p/specs/blob/master/relay/DCUtR.md
+
+#### Open Questions
+
+- Can _Browser_ know upfront its UDP port which it is listening for incoming
+  connections on? Does the browser reuse the UDP port across many WebRTC
+  connections? If that is the case one could connect to any public node, with
+  the remote telling the local node what port it is perceived on.
+
+- Can _Browser_ control the lifecycle of its local TLS certificate, i.e. can
+  _Browser_ use the same TLS certificate for multiple WebRTC connections?
+
+  Yes. For the lifetime of the page, one can generate a certificate once and
+  reuse it across connections. See also
+  https://developer.mozilla.org/en-US/docs/Web/API/RTCPeerConnection/RTCPeerConnection#using_certificates
+
+- Can two _Browsers_ exchange their SDP packets via a third server node using
+  Circuit Relay v2 and DCUtR? Instead of exchanging the original SDP packets,
+  could they exchange their multiaddr and construct the remote's SDP packet
+  based on it?
+
+## Connection Security
+
+While WebRTC offers confidentiality and integrity via TLS, one still needs to
+authenticate the remote peer by its libp2p identity.
+
+After [Connection Establishment](#connection-establishment):
+
+1. _A_ opens a WebRTC datachannel and starts a Noise handshake using _A_'s and
+   _B_'s libp2p identity. See
+   [noise-libp2p](https://github.com/libp2p/specs/tree/master/noise).
+
+2. _A_ and _B_ write their TLS certificate fingerprint on the negotiated Noise
+   channel as multibase encoded multihashes (see [#addressing]).
```

Comment:
This costs us one round trip. The Noise handshake takes one round trip, and then exchanging the fingerprints takes another one.

Since the fingerprints are not private information, we should do something similar to what we do in WebTransport (#404): use Noise's early data. Specifically, both client and server can send their respective messages alongside the first Noise handshake message.

This leads to the question how this should be encoded. The go-to encoding in libp2p would be Protobuf, but I've argued in #429 that it might make sense to use CBOR here.

Reply:
> Since the fingerprints are not private information, we should do something similar to what we do in WebTransport (https://github.com/libp2p/specs/pull/404): use Noise's early data. Specifically, both client and server can send their respective messages alongside the first Noise handshake message.

Agreed, that we should look into Noise's early data mechanism. I don't have enough experience with Noise (yet) to make an informed decision. Input welcome.

> This leads to the question how this should be encoded. The go-to encoding in libp2p would be Protobuf, but I've argued in https://github.com/libp2p/specs/pull/429 that it might make sense to use CBOR here.

For others, see discussion on #429.

Reply:
Moving discussion on latency optimization (not encoding) to https://github.com/libp2p/specs/pull/412/files#r930582351.

Reply:
Resolving here. I suggest discussing CBOR vs Protobuf on #429.